src/lib/c-utils.h

```diff
@@ -69,9 +70,41 @@ static inline
 r_ssize r_ssize_min(r_ssize x, r_ssize y) {
   return (y < x) ? y : x;
 }
+
 static inline
 r_ssize r_ssize_max(r_ssize x, r_ssize y) {
   return (y < x) ? x : y;
 }
 
+static inline
+double r_ssize_as_double(r_ssize x) {
+  if (x > DBL_MAX || x < -DBL_MAX) {
+    r_stop_internal("r_ssize_as_double", "Result can't be represented as `double`.");
+  }
+
+  return (double) x;
+}
+
+static inline
+r_ssize r_double_as_ssize(double x) {
+  x = round(x);
+
+  if (x > R_SSIZE_MAX || x < R_SSIZE_MIN) {
```

Comment:
do you want `x < 0` rather than `< R_SSIZE_MIN ` maybe? i don't know if you want to error on negative sizes here, but maybe:

```c
static inline
r_ssize r_double_as_ssize(double x) {
  if (x > R_SSIZE_MAX || x < 0) {
    r_stop_internal();
  }

  x = floor(x);

  return (r_ssize) x;
}
```

in which case the `floor()` might even be repetitive because of the cast to `(r_ssize)`

Reply:
Since this is specifically a signed type, I think we should allow negative integers. They might be error codes.